### ISSUE TITLE
Odd convention for the tensor product?

### ISSUE THREAD

**Issue by david-pl** [qojulia/QuantumOptics.jl#177]
I was trying to implement the partial transpose and ran into some problems. It took me a while to realize that this was actually due to a convention used throughout the toolbox. Namely, the tensor product is defined in a way that is exactly opposite to what one is used to from standard notations, i.e.

```julia
b = SpinBasis(1//2)
tensor(spinup(b), spindown(b)) 

Ket(dim=4)
  basis: [Spin(1/2) ⊗ Spin(1/2)]
 0.0+0.0im
 0.0+0.0im
 1.0+0.0im
 0.0+0.0im
```
even though
```julia
kron(spinup(b).data, spindown(b).data)

4-element Array{Complex{Float64},1}:
 0.0+0.0im
 1.0+0.0im
 0.0+0.0im
 0.0+0.0im
```
which is the usual way to write a tensor product. This is due the definition of the tensor function in the toolbox, as e.g. for states in https://github.com/qojulia/QuantumOptics.jl/blob/9ab5d4b5dfb6cc40fb02d235e4250f0c9a957f9c/src/states.jl#L90

Since this "switched" tensor product is consistent throughout the toolbox, no issues arise. Is there a reason, though, why it is defined like this? Shouldn't we use the more common convention in order to avoid confusions?

**Comment by david-pl** [comment]
Ok so there is a reason for this convention and it is as follows: `kron(a, b)` basically returns data that is ordered by rows with respect to the `a` and `b`. However, due to Julia's way of storing arrays (column-major order), it is preferable for performance when stepping through the subspaces to have them column-ordered. This can be achieved by simply swapping `a` and `b`, which is the way the tensor product is implemented now.

Changing the tensor product to "normal" ordering has a direct impact on performance. For example, when changing the implementation of `ptrace` according to the "normal" ordering, we see that the speed of the function takes a hit.

![bench-ptrace-operator](https://user-images.githubusercontent.com/18166442/36891666-ac5a3500-1e02-11e8-8a84-1851e14eeafe.png)

Here, d8e9 is the latest master, whereas 17f2 features the "normal" ordering of the tensor product.

For reference: the implementation of `ptrace` can be changed by simply changing the `operators_dense._strides` function to

```julia
function _strides(shape::Vector{Int})
    N = length(shape)
    S = zeros(Int, N)
    S[end] = 1
    for m=1:N-1
        S[N-m] = S[N-m+1]*shape[N-m+1]
    end
    return S
end
```

We should add a comment in the documentation on this.

**Comment by david-pl** [comment]
A paragraph discussing the reason for this ordering is now included in the documentation. It also refers to the function implemented in #207 that allows changing the output.
